src/Features/CSharp/Portable/Copilot/CSharpCopilotCodeFixProvider.cs

```diff
@@ -0,0 +1,111 @@
+{
+    private const string FixPropertyName = "Fix";
+    private const string PromptTitlePropertyName = "PromptTitle";
+
+    private static SyntaxAnnotation WarningAnnotation { get; }
+        = CodeActions.WarningAnnotation.Create(
+            CSharpFeaturesResources.Warning_colon_AI_suggestions_might_be_inaccurate);
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CSharpCopilotCodeFixProvider()
+    {
+    }
+
+    public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+    {
+        var document = context.Document;
+        var cancellationToken = context.CancellationToken;
+
+        var copilotService = document.Project.Solution.Services.GetService<ICopilotCodeAnalysisService>();
+        if (copilotService is null || !copilotService.IsCodeAnalysisOptionEnabled(document))
+            return;
+
+        var isAvailable = await copilotService.IsAvailableAsync(document, cancellationToken).ConfigureAwait(false);
+        if (!isAvailable)
+            return;
+
+        var promptTitles = await copilotService.GetAvailablePromptTitlesAsync(document, cancellationToken).ConfigureAwait(false);
+        if (promptTitles.IsEmpty)
+            return;
+
+        var hasMultiplePrompts = promptTitles.Length > 1;
+
+        // Find the containing method, if any, and also update the fix span to the entire method.
+        // TODO: count location in doc-comment as part of the method.
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        var containingMethod = CSharpSyntaxFacts.Instance.GetContainingMethodDeclaration(root, context.Span.Start, useFullSpan: false);
+        if (containingMethod is not BaseMethodDeclarationSyntax)
+            return;
+
+        foreach (var diagnostic in context.Diagnostics)
+        {
+            Debug.Assert(containingMethod.FullSpan.IntersectsWith(diagnostic.Location.SourceSpan));
+
+            var fix = TryGetFix(document, containingMethod, diagnostic, hasMultiplePrompts);
+            if (fix != null)
+                context.RegisterCodeFix(fix, diagnostic);
+        }
+    }
+
+    private static CodeAction? TryGetFix(
+        Document document,
+        SyntaxNode method,
+        Diagnostic diagnostic,
+        bool hasMultiplePrompts)
+    {
+        if (diagnostic.Properties.TryGetValue(FixPropertyName, out var fix) is false ||
+            diagnostic.Properties.TryGetValue(PromptTitlePropertyName, out var promptTitle) is false)
+        {
+            return null;
+        }
+
+        var title = hasMultiplePrompts
+            ? $"{CSharpFeaturesResources.Apply_fix_from} {promptTitle}"
+            : CSharpFeaturesResources.Apply_Copilot_suggestion;
+
+        return new CopilotDocumentChangeCodeAction(
+            title,
+            createChangedDocument: (_, cancellationToken) => GetFixedDocumentAsync(method, fix!, cancellationToken),
```

Comment:
something is wonky with nrt here.  you should check if fix is null before getting here.